src/interceptor/NodeClientRequest.ts

```diff
@@ -0,0 +1,152 @@
+  public requestId: string | null = null;
+  public ignoredDomains: string[];
+
+  constructor(
+    [url, requestOptions, callback]: NormalizedClientRequestArgs,
+    options: NodeClientOptions
+  ) {
+    super(requestOptions, callback);
+
+    this.requestId = crypto.randomUUID();
+    this.url = url;
+    this.emitter = options.emitter;
+    this.ignoredDomains = options.ignoredDomains ?? [];
+
+    // Set request buffer to null by default so that GET/HEAD requests
+    // without a body wouldn't suddenly get one.
+    this.requestBuffer = null;
+  }
+
+  private writeRequestBodyChunk(
+    chunk: string | Buffer | null,
+    encoding?: BufferEncoding
+  ): void {
+    if (chunk == null) {
+      return;
+    }
+
+    if (this.requestBuffer == null) {
+      this.requestBuffer = Buffer.from([]);
+    }
+
+    const resolvedChunk = Buffer.isBuffer(chunk)
+      ? chunk
+      : Buffer.from(chunk, encoding);
+
+    this.requestBuffer = Buffer.concat([this.requestBuffer, resolvedChunk]);
+  }
+
+  write(...args: ClientRequestWriteArgs): boolean {
+    const [chunk, encoding, callback] = normalizeClientRequestWriteArgs(args);
+
+    // Write each request body chunk to the internal buffer.
+    this.writeRequestBodyChunk(chunk, encoding);
+
+    return super.write(chunk, encoding as BufferEncoding, callback);
+  }
+
+  end(cb?: (() => void) | undefined): this;
+  end(chunk: any, cb?: (() => void) | undefined): this;
+  end(
+    chunk: any,
+    encoding: BufferEncoding,
+    cb?: (() => void) | undefined
+  ): this;
+  end(chunk?: unknown, encoding?: unknown, cb?: unknown): this {
+    if (!this.ignoredDomains.includes(this.url.hostname)) {
+      const requestBody = getArrayBuffer(this.requestBuffer ?? Buffer.from([]));
+      this.emitter.emit(
+        'request',
+        this.toIsomorphicRequest(requestBody),
+        this.requestId
+      );
+    }
+    return super.end(chunk, encoding as BufferEncoding, cb as () => void);
+  }
+
+  emit(event: 'close'): boolean;
+  emit(event: 'drain'): boolean;
+  emit(event: 'error', err: Error): boolean;
+  emit(event: 'finish'): boolean;
+  emit(event: 'pipe', src: Readable): boolean;
+  emit(event: 'unpipe', src: Readable): boolean;
+  emit(event: string | symbol, ...args: any[]): boolean {
+    if (event === 'response') {
+      async function emitResponse(
+        requestId: string,
+        message: IncomingMessage,
+        emitter: EventEmitter
+      ) {
+        const response = args[0] as IncomingMessage;
+        const responseBody = await getIncomingMessageBody(message);
+        emitter.emit(
+          'response',
+          {
+            status: response.statusCode || 200,
+            statusText: response.statusMessage || 'OK',
+            headers: createHeadersFromIncomingHttpHeaders(message.headers),
+            body: responseBody
+          },
+          requestId
+        );
+      }
+
+      if (!this.ignoredDomains.includes(this.url.hostname)) {
```

Comment:
can you hoist this check into a helper function that can also accept an option to ignore internal domains?

Reply:
ignoring internal domains should be default 'true' and I realize the check is imperfect, but I was hoping we can do something like this:

```export function isLocalHostname (hostname: string): boolean {
  if(!hostname) return true;
  const [domain, tld] = hostname.split('.')
  if(!tld) {
    return true
  }
  if(['local'].includes(tld)) {
    return true
  }
  return false;
}
```